[Hub Generated] Review request for Microsoft.HybridCompute to add version preview/2021-01-28-preview

dkirby017 opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/dkirby017/azure-rest-api-specs/dev-hybridcompute-2021-01-28-preview...master/).

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.  **March**
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.  **March**
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [x] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “**WaitForARMFeedback**” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://msazure.visualstudio.com/One/_workitems/create/Scenario?templateId=d1c215ec-5d6a-4cb3-a193-95ddd0a1ba96&ownerId=85bb92fb-21fe-41ca-9309-47f2c5f57515). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

dkirby017 committed:
Adds base for updating Microsoft.HybridCompute from version preview/2020-08-15-preview to version 2021-01-28-preview

dkirby017 committed:
Updates readme

dkirby017 committed:
Updates API version in new specs and examples

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 committed:
update api

dkirby017 commented:
/azp run unifiedPipeline

dkirby017 commented:
@majastrz Following up on your PR comment on the /machines/<machine>/privateLinkScope api .

Given that we want to get the current private link scope's data assigned to a given machine, without needing to know anything more than just the machine resource id, and that there will only ever be at most one, how would you recommend we structure this api?

dkirby017 commented:
Re: the SDK Track2 Validation error. I'm not clear what its actually complaining about :(

zhenglaizhang commented:
/azp run

majastrz reviewed (approved):
Signed off from ARM side.

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "allOf": [
```
do not use inline allOf

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+    },
+    "MachineExtensionProperties": {
+      "properties": {
+        "forceUpdateTag": {
```
What are the possible values for this?  This feels like something where there are know values the user should pass.  This should be documented as a closed or open enum

dkirby017 replied:
I'd like to defer any changes to Extensions for now. these have been in for a long time and several api versions, rather than being added or edited in this api version. I'm not clear why these were approved previously if they're not acceptable now, but i'd like to pass this over to a SME in the area on our team for a follow up PR, rather than blocking the changes this PR was intended for

markcowl replied:
This is mainly a usability concern - how does the user know what values this might contain.  It is ok to defer for now

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+        },
+        "settings": {
+          "type": "object",
+          "description": "Json formatted public settings for the extension."
```
This is under-specification of the settings.  In many languages, there will be no way to serialize or deserialize this value.  You need to have some modelling of these settings

raghushantha replied:
@markcowl  - Settings is an open type that is used by Extension publishers to include properties specific to their Extension. Do you a suggestion on how to better model this?

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+          "type": "object",
+          "description": "Json formatted public settings for the extension."
+        },
+        "protectedSettings": {
```
Again, this is under-modelling.  When youy leave a field as 'object' with no further definition, it makes it essentially impossible for a customer to use this field.

markcowl replied:
Based on the description, there should be a base schema, one for protectedSettings, and one for protectedSettingsFromKeyVault

raghushantha replied:
@markcowl  - Same comment as above. ProtectedSettings is an open type and we don't know how different publishers are going to use this.
this is similar to how CRP models these properties

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+        "provisioningState": {
+          "readOnly": true,
+          "type": "string",
+          "description": "The provisioning state, which only appears in the response."
```
Please provide an open or closed enum with expected values.

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+          "description": "The provisioning state, which only appears in the response."
+        },
+        "instanceView": {
+          "allOf": [
```
DO not use inline all-of, use direct schema reference

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+    },
+    "MachineExtensionUpdateProperties": {
+      "properties": {
+        "forceUpdateTag": {
```
same comment as above:  what are the accepted values?

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+          "type": "string",
+          "description": "The name of the extension handler publisher."
+        },
+        "type": {
```
Given this, with the unspecificed setting below, this feels like it should be a polymorphic type, with this 'type' field determining the schema used for settings, with some settings being universal for all extensions

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+          "type": "string",
+          "description": "Specifies the version of the script handler."
+        },
+        "autoUpgradeMinorVersion": {
```
Advice:  booleans are difficult to interpret.  It is generally clearer to have an enumeration with two clear settings, i.e.:
'MinorVersionUpgradeIfAvailable' | 'DoNotUpgrade'

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+          "type": "boolean",
+          "description": "Indicates whether the extension should use a newer minor version if one is available at deployment time. Once deployed, however, the extension will not upgrade minor versions unless redeployed, even with this property set to true."
+        },
+        "settings": {
```
see comments above on under-specification of these fields, and possibly using polymorphism to provide better modeling for end users

markcowl replied:
Note that extensions have always been difficult to manage because they are generally provided by different teams, but there really needs to be some kind of polymorphic or generric schema here that allows specifying extension properties, or this field is essentially impossible to use.  If your timeine is urgent, we can defer this to a future version, but we should ensure there is an issue filed to look at the modeling of extensions.

markcowl commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+          "description": "Specifies the version of the script handler."
+        },
+        "status": {
+          "properties": {
```
Is this a resource?  If so, it should be modeled as a resource

markcowl replied:
I am asking if MachineExtensionImageView is a resource (the corresponding VirtualMachineInstanceView is)

raghushantha replied:
@markcowl - InstanceView is the structure of the data returned during GET Extensions. The modelling of the data is shown here

dkirby017 committed:
revert changes to package-lock.json

dkirby017 commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+          "Create or Update a Machine": {
+            "$ref": "./examples/Machines_CreateOrUpdate.json"
+          }
+        }
```
it is not

dkirby017 commented on package-lock.json:
```diff
@@ -1,8 +1,8148 @@
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
   "requires": true,
+  "packages": {
```
agreed. no idea why it got changed :)

dkirby017 commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+            "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "name": "resourceGroupName",
```
none are long running

dkirby017 commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "allOf": [
```
these were all in from previous api versions, so I'm unsure where they're actually appropriate. I got the impression from looking into it that we've been using them all wrong, so I removed all of the "allOf" references that only contained a single item. hopefully thats right

markcowl replied:
allOf implies inheritance rather than composition in our generators, so it is appropriate in some cases (models extending Resource), but not in others (places where you want to use a specific type).  In terms of the REST API, these are interchangeable, but they have very different effects for client generation

dkirby017 commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+    "MachineUpdate": {
+      "properties": {
+        "identity": {
+          "x-ms-client-flatten": true,
```
i have no idea! 😄 these were already here, and we don't expect to flatten any of our responses, as far as I'm aware. removed all of them

dkirby017 commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+        "message"
+      ]
+    },
+    "ErrorResponse": {
```
this was here from our first swagger version back in 2019. no idea if the common types existed then. happy to change it.

dkirby017 commented on specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2021-01-28-preview/HybridCompute.json:
```diff
@@ -0,0 +1,1205 @@
+        "error"
+      ]
+    },
+    "Identity": {
```
again, it's been there since 2019, but happy to change it

dkirby017 committed:
address a bunch of comments

dkirby017 committed:
run prettier and fix errorresponse reference

dkirby017 committed:
update parameter name in examples

dkirby017 committed:
fix parameter names

dkirby017 committed:
got lots of errors after removing the allOfs so following adps example and using them at the top level of the definitions

dkirby017 committed:
fix some lint errors

dkirby017 commented:
@markcowl I've addressed most of your comments and left a note on the extension section. I'm not sure what it's complaining about at this point, but please take a look as soon as you can.

dkirby017 committed:
use common errorDetail definition

dkirby017 commented:
@markcowl everything looks happier after my latest update

markcowl commented:
@dkirby017  See my comments on the unresolved comments above.  Underspecifying extensions is a long-standing problem, so if this is urgent, we can file an issue and defer until later.  I would hate to do this with settings/protectSettings, but am open to treating these the same way if your team will fix the problem in an upcoming api-version

dkirby017 commented:
@markcowl it is urgent that we get this new api in. It's blocking an upcoming preview of our private link support. I've filed https://msazure.visualstudio.com/One/_workitems/edit/9398639 to address your comments on the extensions and we'll get those in in the next update :)
